server/routerlicious/packages/routerlicious-base/src/nexus/ordererManager.ts

```diff
@@ -0,0 +1,228 @@
+	private readonly ordererConnectionCloseTimeoutMap = new Map<string, NodeJS.Timeout>();
+
+	private readonly options: IOrdererManagerOptions;
+
+	constructor(
+		private readonly globalDbEnabled: boolean,
+		private readonly ordererUrl: string,
+		private readonly tenantManager: ITenantManager,
+		private readonly localOrderManager: LocalOrderManager,
+		private readonly kafkaFactory: KafkaOrdererFactory,
+		options: Partial<IOrdererManagerOptions>,
+	) {
+		this.options = { ...DefaultOrdererManagerOptions, ...options };
+	}
+
+	private startCleanupTimer(tenantId: string, documentId: string): void {
+		// Make sure no timer is running already.
+		this.stopCleanupTimer(tenantId, documentId);
+		const ordererId = this.getOrdererConnectionMapKey(tenantId, documentId);
+		const lumberBaseProperties = getLumberBaseProperties(documentId, tenantId);
+		const cleanupOrdererConnectionTimeout = setTimeout(() => {
+			this.cleanupOrdererConnection(tenantId, documentId)
+				.then(() => {
+					Lumberjack.info(`Successfully cleaned up orderer connection`, {
+						...lumberBaseProperties,
+						remainingConnections: this.ordererConnectionCountMap.size,
+					});
+				})
+				.catch((error) => {
+					Lumberjack.error(
+						`Failed to cleanup orderer connection`,
+						lumberBaseProperties,
+						error,
+					);
+				});
+		}, this.options.ordererConnectionCleanupTimeoutMs);
+		// Cleanup the connection after some time to avoid frequent connection creation/destruction
+		this.ordererConnectionCloseTimeoutMap.set(ordererId, cleanupOrdererConnectionTimeout);
+	}
+	private stopCleanupTimer(tenantId: string, documentId: string): void {
+		const ordererId = this.getOrdererConnectionMapKey(tenantId, documentId);
+		if (this.ordererConnectionCloseTimeoutMap.has(ordererId)) {
+			clearTimeout(this.ordererConnectionCloseTimeoutMap.get(ordererId));
+		}
+	}
+
+	public async getOrderer(tenantId: string, documentId: string): Promise<IOrderer> {
+		const ordererConnectionType = await this.getOrdererConnectionType(tenantId, documentId);
+
+		// Clear any existing connection timeout because a new connection was added.
+		this.stopCleanupTimer(tenantId, documentId);
+
+		if (ordererConnectionType === "local") {
+			Lumberjack.info(`Using local orderer`, getLumberBaseProperties(documentId, tenantId));
+			const localOrderer = await this.localOrderManager.get(tenantId, documentId);
+
+			this.updateOrdererConnectionCount(tenantId, documentId, "increment", "local").catch(
+				(error) => {
+					Lumberjack.error(
+						`Failed to update local orderer connection count`,
+						{
+							documentId,
+							tenantId,
+						},
+						error,
+					);
+				},
+			);
+
+			return localOrderer;
+		}
+
+		Lumberjack.info(`Using Kafka orderer`, getLumberBaseProperties(documentId, tenantId));
+		const kafkaOrderer = await this.kafkaFactory.create(tenantId, documentId);
+
+		this.updateOrdererConnectionCount(tenantId, documentId, "increment", "kafka").catch(
```

Comment:
`await` for these update cont calls?

Reply:
For this, I specifically wanted it to be fire-and-forget to make sure this work doesn't slow down ConnectDocument flow. If `ordererType` isn't in the local Map for some reason, we would have to fetch from the DB which is slower. This should never happen, but it felt safest from a perf perspective to make it fire/forget. In ideal scenario, it will happen on the next tick, in worst scenario it will not slow down connect.